File: ejemplos/factura_electronica.c
/* 
 * Ejemplo de Uso de Biblioteca LibPyAfipWs (.DLL / .so)
 * con Web Service Autenticación / Factura Electrónica AFIP
 * 2013 (C) Mariano Reingart <reingart@gmail.com>
 * Licencia: GPLv3
 * Requerimientos: scripts wsaa.py y libpyafipws.h / libpyafipws.c
 * Documentacion: 
 *  http://www.sistemasagiles.com.ar/trac/wiki/LibPyAfipWs
 *  http://www.sistemasagiles.com.ar/trac/wiki/ManualPyAfipWs
 */

#include "libpyafipws.h"

#if defined(__GNUC__)
#include <stdio.h>
#include <stdlib.h>
#endif

int main(int argc, char *argv[]) {
  BSTR tra, cms, ta;
  void *wsfev1;
  BSTR ret;
  bool ok;
  long nro;
  
  /* prueba generica, el valor devuelto no debería ser nulo */
  ret = test();
  printf("%s\n", ret);
  PYAFIPWS_Free(ret);
  if (ret == NULL) exit(1);
  
  /* Generar ticket de requerimiento de acceso */
  tra = WSAA_CreateTRA("wsfe", 999);
  printf("TRA:\n%s\n", tra);
  /* Firmar criptograficamente el mensaje */
  cms = WSAA_SignTRA((char*) tra, "reingart.crt", "reingart.key");
  printf("CMS:\n%s\n", cms);  
  /* Llamar al webservice y obtener el ticket de acceso */
  ta = WSAA_LoginCMS((char*) cms);
  printf("TA:\n%s\n", ta);
  
  /* Crear una objeto WSFEv1 (interfaz webservice factura electronica) */
  wsfev1 = PYAFIPWS_CreateObject("wsfev1", "WSFEv1");
  printf("crear wsfev1: %p\n", wsfev1);    /* si funcionó ok, no debe ser NULL! */  
  
  /* conectar al webservice (para produccion cambiar URL) */
  ok = WSFEv1_Conectar(wsfev1, "", "", "");
  printf("concetar: %s\n", ok ? "true" : "false");
  
  /* obtener datos genericos de la interfaz (version y ruta de instalación) */
  ret = PYAFIPWS_Get(wsfev1, "Version");
  printf("wsfev1 Version: %s\n", ret);
  free(ret);
  ret = PYAFIPWS_Get(wsfev1, "InstallDir");
  printf("wsfev1 InstallDir: %s\n", ret);
  free(ret);
  
  /* obtener el estado de los servidores (llama al ws) */
  ok = WSFEv1_Dummy(wsfev1);
  printf("llamar a dummy: %s\n", ok ? "true" : "false");
  /* obtener los atributos devueltos por AFIP */
  ret = PYAFIPWS_Get(wsfev1, "AppServerStatus");
  printf("dummy AppServerStatus: %s\n", ret);
  free(ret);
  ret = PYAFIPWS_Get(wsfev1, "DbServerStatus");
  printf("dummy DbServerStatus: %s\n", ret);
  free(ret);
  ret = PYAFIPWS_Get(wsfev1, "AuthServerStatus");
  printf("dummy AuthServerStatus: %s\n", ret);
  free(ret);

  /* establezco los datos para operar el webservice */
  ok = PYAFIPWS_Set(wsfev1, "Cuit", "20267565393");
  ok = WSFEv1_SetTicketAcceso(wsfev1, (char*) ta);  /* devuelto por WSAA_LoginCMS */

  /* obtengo el ultimo numero de comprobante generado */
  nro = WSFEv1_CompUltimoAutorizado(wsfev1, "1", "1");
  printf("ultimo comprobante: %ld\n", nro);

  /* destruir el objeto */
  PYAFIPWS_DestroyObject(wsfev1);  


  /* liberar la memoria adquirida para los valores devueltos de WSAA */
  PYAFIPWS_Free(ta);
  PYAFIPWS_Free(cms);
  PYAFIPWS_Free(tra);

  return 0;
}

File: ejemplos/FacturaElectronica.java
/* Ejemplo de Uso de Interfaz PyAfipWs para JAVA (componentes DLL en Windows)
   con Web Service Autenticación / Factura Electrónica AFIP (mercado interno)
   2014 (C) Mariano Reingart <reingart@gmail.com> Licencia: GPLv3
   
   Requerimientos: 
    * wsaa.py y wsfev1.py registrados (último instalador PyAfipWs homologación)
   
   Dependencias: 
    * JACOB - Java COM Bridge: http://sourceforge.net/projects/jacob-project/
   
   IMPORTANTE: 
    * Renombrar jacob-1.18-M2-x64.dll o jacob-1.18-M2-x86.dll -> jacob.dll
    * Mover jacob.dll al directorio windows\system o junto a esta clase
    * Agregar jacob.jar al CLASSPATH, ej SET CLASSPATH=Z:\ruta\jacob.jar;.
   
   Documentacion: 
    http://www.sistemasagiles.com.ar/trac/wiki/PyAfipWs
    http://www.sistemasagiles.com.ar/trac/wiki/ManualPyAfipWs
*/

import com.jacob.activeX.ActiveXComponent;
import com.jacob.com.Dispatch;
import com.jacob.com.LibraryLoader;
import com.jacob.com.Variant;

public class FacturaElectronica {

    public static void main(String[] args) {
        try {

            LibraryLoader.loadJacobLibrary();
            
            /* Crear objeto WSAA: Web Service de Autenticación y Autorización */
            ActiveXComponent wsaa = new ActiveXComponent("WSAA");
            
            System.out.println(Dispatch.get(wsaa, "InstallDir").toString() + 
                               Dispatch.get(wsaa, "Version").toString()
                              );
                        
            /* Solicitar Ticket de Acceso a AFIP (cambiar URL producción) */
            String wsdl = "https://wsaahomo.afip.gov.ar/ws/services/LoginCms";
            String userdir = System.getProperty("user.dir");
            Dispatch.call(wsaa, "Autenticar", 
                                new Variant("wsfe"), 
                                new Variant(userdir + "/reingart.crt"), 
                                new Variant(userdir + "/reingart.key"), 
                                new Variant(wsdl));
            String excepcion =  Dispatch.get(wsaa, "Excepcion").toString();
            System.out.println("Excepcion: " + excepcion);
            String token = Dispatch.get(wsaa, "Token").toString();
            String sign = Dispatch.get(wsaa, "Sign").toString();
            System.out.println("Token: " + token +  "Sign: " + sign);
            
            /* Instanciar WSFEv1: WebService de Factura Electrónica version 1 */

            ActiveXComponent wsfev1 = new ActiveXComponent("WSFEv1");

            /* Establecer parametros de uso: */
            Dispatch.put(wsfev1, "Cuit", new Variant("20267565393"));
            Dispatch.put(wsfev1, "Token", new Variant(token));
            Dispatch.put(wsfev1, "Sign", new Variant(sign));

            /* Conectar al websrvice (cambiar URL para producción) */
            wsdl = "https://wswhomo.afip.gov.ar/wsfev1/service.asmx?WSDL";
            Dispatch.call(wsfev1, "Conectar", 
                                  new Variant(""), 
                                  new Variant(wsdl));

            /* Consultar último comprobante autorizado en AFIP */
            String tipo_cbte = "1";
            String pto_vta = "1";
            Variant ult = Dispatch.call(wsfev1, "CompUltimoAutorizado", 
                                                new Variant(tipo_cbte), 
                                                new Variant(pto_vta));
            System.out.println("Ultimo comprobante: " + ult.toString());
            excepcion =  Dispatch.get(wsfev1, "Excepcion").toString();
            System.out.println("Excepcion: " + excepcion);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
